skorch/net.py

```diff
@@ -700,6 +904,51 @@ def get_train_step_accumulator(self):
         """
         return FirstStepAccumulator()
 
+    def _zero_grad_optimizer(self, set_to_none=None):
+        """Zero out the gradient of all optimizers.
+
+        Parameters
+        ----------
+        set_to_none : bool or None (default=None)
+          Whether to zero out gradients (default) or to set them to None by
+          passing True. Note that since this option is only available starting
+          from PyTorch 1.7, it is ignored by default (i.e. when its value is
+          None). For skorch to pass this value to the ``zero_grad`` call,
+          override this method and set the value to True or False.
+
+          The advantages and disadvantages of setting this value to True are
+          discussed here:
+
+          https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer.zero_grad
+
+        """
+        for name in self.optimizers_:
+            optimizer = getattr(self, name + '_')
+            if set_to_none is None:
+                optimizer.zero_grad()
+            else:
+                optimizer.zero_grad(set_to_none=set_to_none)
+
+    def _step_optimizer(self, step_fn):
+        """Perform a ``step`` call on all optimizers.
+
+        Parameters
+        ----------
+        step_fn : callable or None
+          If None, just call ``optimizer.step()`` without arguments. Else, this
+          will be passed as the training step closure to the optimizer(s). Note
+          that this could lead to the function being called multiple times. If
+          more fine-grained control is desired instead, please override the
+          :meth:`.train_step` method.
+
+        """
+        for name in self.optimizers_:
+            optimizer = getattr(self, name + '_')
```

Comment:
Using string mangling to get the optimizers or modules is an okay approach.

I am thinking of using `WeakValueDictionary` to hold a weak reference could be better. It could even keep track of itself to avoid the whole `_unregister_attribute:

```python
from weakref import WeakValueDictionary

class A:
    pass

a_obj = A()

my_dict = WeakValueDictionary() 
my_dict["hello"] = a_obj

# object is in dict
assert "hello" in my_dict

# Remove object
del a_obj

# object not in dict
assert "hello" not in my_dict
```

If we use this `WeakValueDictionary`, then `modules_` can be a `@property` returning `self.modules_weak_dict_.keys()`.

I do not have a strong opinion. (`WeakValueDictionary` is not a well know python object)